File: Wire-iOS/Sources/Components/AudioPlayer/AudioTrackCell.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


@import PureLayout;
@import WireExtensionComponents;

#import "AudioTrackCell.h"
#import "AudioTrackView.h"
#import "IconButton.h"
#import "AudioTrack.h"


@implementation AudioTrackCell

- (void)prepareForReuse
{
    [self.audioTrackView.playPauseButton setIcon:ZetaIconTypePlay withSize:ZetaIconSizeLarge forState:UIControlStateNormal];
    self.audioTrackView.progress = 0;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    
    if (self) {
        _audioTrackView = [[AudioTrackView alloc] initForAutoLayout];
        [self.contentView addSubview:self.audioTrackView];
        [self.audioTrackView.playPauseButton addTarget:self action:@selector(playPause:) forControlEvents:UIControlEventTouchUpInside];
        [self.audioTrackView autoPinEdgesToSuperviewEdgesWithInsets:UIEdgeInsetsZero];
    }
    
    return self;
}

- (void)setAudioTrack:(id<AudioTrack>)audioTrack
{
    _audioTrack = audioTrack;
    
    self.artworkObserver = [KeyValueObserver observeObject:self.audioTrack keyPath:@"artwork" target:self selector:@selector(artworkChanged:) options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew];
    
    if (self.audioTrack.artwork == nil) {
        [self.audioTrack fetchArtwork];
    } else {
        [self updateArtwork];
    }
    
    self.audioTrackView.failedToLoad = self.audioTrack.failedToLoad || self.audioTrack == nil;
}

- (void)updateArtwork
{
    self.audioTrackView.artworkImageView.image = self.audioTrack.artwork;
}

- (void)artworkChanged:(NSDictionary *)change
{
    [self updateArtwork];
}

- (IBAction)playPause:(id)sender
{
    [self.delegate audioTrackCell:self didPlayPauseTrack:self.audioTrack];
}

@end

File: Wire-iOS/Sources/Components/AudioPlayer/AudioTrackCell.swift
//
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//
import UIKit

@objc protocol AudioTrackCellDelegate: NSObjectProtocol {
    func audioTrackCell(_ cell: AudioTrackCell?, didPlayPauseTrack audioTrack: AudioTrack?)
}

@objcMembers
final class AudioTrackCell: UICollectionViewCell {
    var audioTrack: AudioTrack? {
        didSet {
            artworkObserver = KeyValueObserver.observe(self.audioTrack, keyPath: "artwork", target: self, selector: #selector(self.artworkChanged(_:)), options: [.initial, .new])

            if audioTrack?.artwork == nil {
                audioTrack?.fetchArtwork()
            } else {
                updateArtwork()
            }

            audioTrackView.failedToLoad = self.audioTrack?.failedToLoad ?? false || self.audioTrack == nil

        }
    }

    let audioTrackView = AudioTrackView()
    var artworkObserver: NSObject?
    weak var delegate: AudioTrackCellDelegate?

    override func prepareForReuse() {
        audioTrackView.playPauseButton?.setIcon(.play, with: .large, for: .normal)
        
        audioTrackView.progress = 0
    }

    override init(frame: CGRect) {
        super.init(frame: frame)

        contentView.addSubview(audioTrackView)
        audioTrackView.playPauseButton.addTarget(self, action: #selector(self.playPause(_:)), for: .touchUpInside)

        audioTrackView.translatesAutoresizingMaskIntoConstraints = false
        audioTrackView.fitInSuperview()
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    func updateArtwork() {
        audioTrackView.artworkImageView.image = audioTrack?.artwork
    }

    func artworkChanged(_ change: [AnyHashable : Any]?) {
        updateArtwork()
    }

    func playPause(_ sender: Any) {
        delegate?.audioTrackCell(self, didPlayPauseTrack: audioTrack)
    }

}
